# 4.5: Data Consistency Checks Practice

## Importing Libraries and Data

In [2]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [3]:
path = r'/Users/leo/Desktop/Career Foundry/Achievement 4/Instacart Basket Analysis 30:10:25'

In [4]:
# importing data set products.csv
df_prods = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'products.csv'),index_col = False)

In [25]:
# importing data set Orders_Wrangled.csv
df_ords = pd.read_csv(os.path.join(path, '02 - Data', 'Prepared Data', 'Orders_Wrangled.csv'),index_col = False)

# Mixed Data Types 

## Creating small test data frame

In [6]:
df_test = pd.DataFrame()

In [7]:
df_test['mix']=['a','b',1,True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

# Function for checking mixed type

In [9]:
# check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type)!=df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    print (col)

mix


/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_42122/1831855508.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type)!=df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [10]:
# convert a column’s data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [11]:
# Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# create new data frame with columns with missing values
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
# See df_prods rows/columns
df_prods.shape

(49693, 5)

# Ceating new Data frame without missing values

In [15]:
# Creating new Data Frame df_prods_clean which does NOT have missing values (False)
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

# Finding Duplicate Values

In [17]:
## Creates new 'df_dups' based on duplicate values it finds in df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Addressing Duplicates

In [19]:
# Check columns and rows
df_prods_clean.shape

(49677, 5)

In [20]:
# Create new Data frame that doesn't include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
# Check to see duplicate values have been dropped
df_prods_clean_no_dups.shape

(49672, 5)

# Exporting Data 

In [22]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 - Data','Prepared Data', 'products_checked.csv'))

# TASK 4.5

## 2

In [23]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## The maximum order number is 100

## days_since_prior_order: The maximum is 30, which seems reasonable, but there are 0 values—likely representing first-time orders.

## order_dow = day of the week. 0 represents monday 6 represents sunday. Giving us 7 days of the week in total.

# 3 Check for mixed data types in df_ords

In [27]:
# check for mixed data types
for col in df_ords.columns.tolist():
    weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    print(col)

/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_42122/1049942620.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)


Unnamed: 0


/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_42122/1049942620.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)


order_id


/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_42122/1049942620.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)


user_id


/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_42122/1049942620.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)


eval_set


/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_42122/1049942620.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)


order_number


/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_42122/1049942620.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)


order_dow


/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_42122/1049942620.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)


order_hour_of_day


/var/folders/19/p3csqgws1vqgpn31c7f5fcqh0000gn/T/ipykernel_42122/1049942620.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird=(df_ords[[col]].applymap(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)


days_since_prior_order


# Was told to use Map instead of applymap 

In [28]:
# check for mixed data types
for col in df_ords.columns.tolist():
    weird=(df_ords[[col]].map(type)!=df_ords[[col]].iloc[0].apply(type)).any(axis=1)
    print(col)

Unnamed: 0
order_id
user_id
eval_set
order_number
order_dow
order_hour_of_day
days_since_prior_order


## 0 mixed data types found

# 4

## No mixed data types found in df_ords

# 5 Run a check for missing values in your df_ords dataframe.

In [30]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## Explenation: There are 206209 missing values in the 'days_since_prior_order column. These missing values could represent first time customers as there is not historical data. Since we cannot impute these values and we can't remove these values, we could implement a flag system as shared in the example with how banks use flag systems.

# 6 Address the missing values using an appropriate method.

## first I want to check if the missing values have an order in the order_number column. 

In [32]:
# making a missing values data frame for df_ords column days_since_prior_order
df_dspo_nan = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [33]:
df_dspo_nan

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
11               11   2168274        2    prior             1          2   
26               26   1374495        3    prior             1          1   
39               39   3343014        4    prior             1          6   
45               45   2717275        5    prior             1          3   
...             ...       ...      ...      ...           ...        ...   
3420930     3420930    969311   206205    prior             1          4   
3420934     3420934   3189322   206206    prior             1          3   
3421002     3421002   2166133   206207    prior             1          6   
3421019     3421019   2227043   206208    prior             1          1   
3421069     3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 8 columns]

In [35]:
# now I will make a data frame showing the number of rows with first orders as a value of 1.
df_ords_first_order = df_ords[df_ords['order_number']==1]

In [36]:
df_ords_first_order

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
11               11   2168274        2    prior             1          2   
26               26   1374495        3    prior             1          1   
39               39   3343014        4    prior             1          6   
45               45   2717275        5    prior             1          3   
...             ...       ...      ...      ...           ...        ...   
3420930     3420930    969311   206205    prior             1          4   
3420934     3420934   3189322   206206    prior             1          3   
3421002     3421002   2166133   206207    prior             1          6   
3421019     3421019   2227043   206208    prior             1          1   
3421069     3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 8 columns]

## Already from this we can see that we get the same count of rows in both data frames. Which reinforces the idea that we need these values and the appropriate solution would be usinf a flag column (True/Flase).

In [37]:
# Made a new column: First order (True/False)
df_ords['First_order'] = df_ords['days_since_prior_order'].isna()

In [38]:
df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329        1    prior             1          2   
1                 1   2398795        1    prior             2          3   
2                 2    473747        1    prior             3          3   
3                 3   2254736        1    prior             4          4   
4                 4    431534        1    prior             5          4   
...             ...       ...      ...      ...           ...        ...   
3421078     3421078   2266710   206209    prior            10          5   
3421079     3421079   1854736   206209    prior            11          4   
3421080     3421080    626363   206209    prior            12          1   
3421081     3421081   2977660   206209    prior            13          1   
3421082     3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  First_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 9 columns]

# 7 Run a check for duplicate values in your df_ords data.

In [39]:
# looking for duplicate values
df_ords_dups = df_ords[df_ords.duplicated()]

In [40]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order, First_order]
Index: []

## There are no duplicates in df_ords

# 8 Export data frames 

In [42]:
df_ords.to_csv(os.path.join(path, '02 - Data', 'Prepared Data', 'orders_cleaned_4.5.csv'))

In [43]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 - Data','Prepared Data', 'products_cleaned_4.5.csv'))

## Added 4.5 at the end just to keep track of where I have made chnages now that i am half way through the achievement.